File: src/test/java/com/digitalpebble/ginkgo/ConfigTest.java
// SPDX-License-Identifier: Apache-2.0

package com.digitalpebble.ginkgo;

import org.junit.jupiter.api.Test;

import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class ConfigTest {

    @Test
    void loadsSuccessfully() {
        Config config = Config.getInstance();
        assertNotNull(config);
    }

    @Test
    void gridCarbonIntensity() {
        assertEquals(352, Config.getInstance().getGridCarbonIntensity(), 1e-9);
    }

    @Test
    void pue() {
        assertEquals(1.16, Config.getInstance().getPue(), 1e-9);
    }

    @Test
    void runnerPowerConsumptionContainsAllRunners() {
        Map<String, Double> runners = Config.getInstance().getRunnerPowerConsumption();
        assertEquals(6, runners.size());
        assertEquals(65, runners.get("ubuntu"), 1e-9);
        assertEquals(45, runners.get("ubuntu-arm"), 1e-9);
        assertEquals(75, runners.get("windows"), 1e-9);
        assertEquals(55, runners.get("windows-arm"), 1e-9);
        assertEquals(40, runners.get("macos"), 1e-9);
        assertEquals(85, runners.get("macos-13"), 1e-9);
    }
}

File: tests/config_test.rs
use ginkgo::config::Config;

fn assert_close(label: &str, actual: f64, expected: f64) {
    let diff = (actual - expected).abs();
    assert!(
        diff < 1e-9,
        "{label}: expected {expected:.9}, got {actual:.9} (diff: {diff:.9})"
    );
}

#[test]
fn loads_successfully() {
    let config = Config::instance();
    assert!(config.grid_carbon_intensity > 0.0);
}

#[test]
fn grid_carbon_intensity() {
    assert_close(
        "grid_carbon_intensity",
        Config::instance().grid_carbon_intensity,
        352.0,
    );
}

#[test]
fn pue() {
    assert_close("pue", Config::instance().pue, 1.16);
}

#[test]
fn runner_power_consumption_contains_all_runners() {
    let runners = &Config::instance().runner_power_consumption;
    assert_eq!(
        runners.len(),
        7,
        "expected 7 runner types, got {}",
        runners.len()
    );

    let cases = [
        ("ubuntu", 4.315),
        ("ubuntu-slim", 1.078),
        ("ubuntu-arm", 2.157),
        ("windows", 4.746),
        ("windows-arm", 2.372),
        ("macos", -1.0),
        ("macos-13", -1.0),
    ];

    for (name, expected) in cases {
        let actual = *runners
            .get(name)
            .unwrap_or_else(|| panic!("runner '{name}' missing from config"));
        assert_close(&format!("runners[{name}]"), actual, expected);
    }
}
